### PR TITLE
Specify Usage in README.md.

### PR DESCRIPTION
**Pull request by moolitayer**

### CODE REVIEW

**Comment by moolitayer** [comment]
cc @zgalor @jhernand @elad661 @yaacov @cben this is a good time to suggest interface changes if you have any ideas (They won't be done here tough).

File: README.md [hunk 1]
```diff
@@ -11,6 +11,103 @@ go get github.com/moolitayer/golang/glog # Dependency
...
+```
+# Uses the builder pattern:
+connection, err := awx.NewConnectionBuilder().
+  Url("http://awx.example.com/api").
+  Username(username).
+  Password(password).
+  Token("TOKEN").
+  Bearer("BEARER").
+  CAFile("/etc/pki/tls/cert.pem").
+  Insecure(insecure).
```
Comment (elad661) [inline README.md:30]: It'd probably be a good idea to mention which of these are optional and which are mandatory.

File: README.md [hunk 2]
```diff
@@ -11,6 +11,103 @@ go get github.com/moolitayer/golang/glog # Dependency
...
+- Jobs
+- Job Templates
+
+### Retrieving resources
+```
+projectsResource := connection.Projects()
+
+// Get a list of all Projects.
+getProjectsRequest := projectsResource.Get()
+getProjectsResponse, err := getProjectsRequest.Send()
```
Comment (elad661) [inline README.md:64]: I don't particularly like the `.Send()` pattern, but I see it's used for `.Filter()`... How do other Go projects do this? is this a common pattern?

File: README.md [hunk 3]
```diff
@@ -11,6 +11,103 @@ go get github.com/moolitayer/golang/glog # Dependency
...
+  CAFile("/etc/pki/tls/cert.pem").
+  Insecure(insecure).
+  Proxy(http://myproxy.example.com).
+  Build()                                      # Create the client
+if err != nil {
+  panic(err)
+}
+defer connection.Close()                      # Don't forget to close the connection!
+```
+`Url()` points at an AWX server's root API endpoint.
```
Comment (elad661) [inline README.md]: Might be worth it to mention that this needs to include `/api`

File: README.md [hunk 4]
```diff
@@ -11,6 +11,103 @@ go get github.com/moolitayer/golang/glog # Dependency
...
+`Proxy()` specifies A proxy server to use for all outgoing connection to the AWX server.
```
Comment (elad661) [inline README.md]: lowercase a.

File: README.md [hunk 5]
```diff
@@ -11,6 +11,103 @@ go get github.com/moolitayer/golang/glog # Dependency
...
+#### Authentication
```
Comment (elad661) [inline README.md]: Maybe add "use one of the following"?

File: README.md [hunk 6]
```diff
@@ -11,6 +11,103 @@ go get github.com/moolitayer/golang/glog # Dependency
...
+`Username()` and `Password()` specify Basic Auth for AWX API server.
+`Token()` uses the authtoken/ endpoint and works with AWX XXX and Ansible tower < 3.3.
+`Bearer()` uses OAuth2 and works since AWX XXX and Ansible Tower > 3.3.
+
+#### TLS
+`CAFile()` specifies a trusted PEM encoded CA certificates used to verify the AWX server.
```
Comment (elad661) [inline README.md]: Nit: this specifies the path to the file, not the PEM-encoded text. Obvious from the name, but the description doesn't much. Also maybe mention that the host trust store will be used if no CA is provided.

File: README.md [hunk 7]
```diff
@@ -11,6 +11,103 @@ go get github.com/moolitayer/golang/glog # Dependency
...
+if err != nil {
+  panic(err)
+}
+defer connection.Close()                      # Don't forget to close the connection!
+```
+`Url()` points at an AWX server's root API endpoint.
+`Proxy()` specifies A proxy server to use for all outgoing connection to the AWX server.
+#### Authentication
+`Username()` and `Password()` specify Basic Auth for AWX API server.
+`Token()` uses the authtoken/ endpoint and works with AWX XXX and Ansible tower < 3.3.
```
Comment (elad661) [inline README.md]: "AWX XXX"? If we don't know the version number it's probably best to reword this to something like "works with older AWX versions"

File: README.md [hunk 8]
```diff
@@ -11,6 +11,103 @@ go get github.com/moolitayer/golang/glog # Dependency
...
+`Bearer()` uses OAuth2 and works since AWX XXX and Ansible Tower > 3.3.
+
+#### TLS
+`CAFile()` specifies a trusted PEM encoded CA certificates used to verify the AWX server.
+`CAFile()` can be used multiple times to specify a list of files.
+`Insecure(true)` can be specified to ignore TLS verification.
```
Comment (elad661) [inline README.md]: I'd use "disable" instead of "ignore"

File: README.md [hunk 9]
```diff
@@ -11,6 +11,103 @@ go get github.com/moolitayer/golang/glog # Dependency
...
+getProjectResponse, err := projectResource.Get().Send()
+```
+
+#### Launching a Job from a Template
+```
+// Launch Job Template with id=8
+connection.JobTemplates().Id(8).Launch()
+launchResource := connection.JobTemplates().Id(templateId).Launch()
+response, err := launchResource.Post().
+  ExtraVars("{\"awx_environment\": \"staging\"}").
```
Comment (elad661) [inline README.md]: Should we support passing ExtraVars as a map/struct?

File: README.md [hunk 10]
```diff
@@ -11,6 +11,103 @@ go get github.com/moolitayer/golang/glog # Dependency
...
+  Limit("master.example.com").
+  Send()
+if err != nil {
+  return err
+}
+```
+`ExtraVars()` Specify a JSON objects of extra variable to pass with the job templates
```
Comment (elad661) [inline README.md]: This probably should be JSON object, not objects (singular). Otherwise, remove "a".

**Review by elad661 — COMMENTED** [review]
Looks good!

Most of my feedback is phrasing nitpicks, but I have one or two interface questions as well.

File: README.md (hunk 9 above)
Comment (cben) [reply README.md]: AFAICT it already does, takes a map not a string:
https://github.com/moolitayer/awx-client-go/blob/78fe047d29a6a1fdcf7f5f0d9487706a1784b2e2/awx/job_template_launch_resource.go#L86 ?

btw, consider documenting `ExtraVar` too.

File: README.md [hunk 11]
```diff
@@ -11,6 +11,103 @@ go get github.com/moolitayer/golang/glog # Dependency
 go get github.com/moolitayer/awx-client-go/awx
 ```
 
+## Usage
+### import
+```
```
Comment (jhernand) [inline README.md]: If you use `` ```go`` the result will be rendered using syntax highlight, which looks nicer.

File: README.md [hunk 12]
```diff
@@ -11,6 +11,103 @@ go get github.com/moolitayer/golang/glog # Dependency
...
+import 	"github.com/moolitayer/awx-client-go/awx"
+```
+
+### Creating a connection:
+```
+# Uses the builder pattern:
+connection, err := awx.NewConnectionBuilder().
+  Url("http://awx.example.com/api").
```
Comment (jhernand) [inline README.md]: We should probably rename `Url` to `URL`, to stick to the Go naming conventions, but not in this patch.

File: README.md [hunk 13]
```diff
@@ -11,6 +11,103 @@ go get github.com/moolitayer/golang/glog # Dependency
...
+  fmt.Printf("%d: %s - %s\n", project.Id(), project.Name(), project.SCMURL())
+}
+```
+#### Filtering
+User `Filter()` on a request to filter lists:
+```
+projectsResource := connection.Projects()
+
+// Get a list of all Projects using git SCM.
+getProjectsResponse, err := getProjectsRequest.Filter("scm_type", "git").Send()
```
Comment (jhernand) [inline README.md]: In this example the creation of the request object is missing. I'd suggest to include it:

```go
projectsResource := connection.Projects()

// Get a list of all projects using git SCM:
getProjectsResponse, err := projectsResource.Get().
  Filter("scm_type", "git").
  Send()
```

File: README.md [hunk 14]
```diff
@@ -11,6 +11,103 @@ go get github.com/moolitayer/golang/glog # Dependency
...
+
+```
+#### Retrieving resource by id
+Use `id()` on a resource list to get a single resource
```
Comment (jhernand) [inline README.md]: I'd suggest `Id(...)`, as the `Id` method needs to receive the actual identifier as parameter.

File: README.md [hunk 15]
```diff
@@ -11,6 +11,103 @@ go get github.com/moolitayer/golang/glog # Dependency
...
+```
+projectsResource := connection.Projects()
+
+// Get project with id=4
```
Comment (jhernand) [inline README.md]: Note that this does *not* return the project with identifier 4, it rather returns the resource that manages project with identifier 4. The difference is important: the resource is not the object, but the thing that manages it, and that provides the mechanisms to retrieve, update and delete it.

File: README.md [hunk 16]
```diff
@@ -11,6 +11,103 @@ go get github.com/moolitayer/golang/glog # Dependency
...
+projectResource := connection.Projects().Id(4)
+
+// Send the request to retrieve the project:
+getProjectResponse, err := projectResource.Get().Send()
+```
+
+#### Launching a Job from a Template
+```
+// Launch Job Template with id=8
+connection.JobTemplates().Id(8).Launch()
```
Comment (jhernand) [inline README.md]: This ^ line is redundant.

File: README.md [hunk 17]
```diff
@@ -11,6 +11,103 @@ go get github.com/moolitayer/golang/glog # Dependency
...
+launchResource := connection.JobTemplates().Id(templateId).Launch()
```
Comment (jhernand) [inline README.md]: s/templateId/8/

File: README.md (hunk 12 above)
Comment (moolitayer) [reply README.md]: Thanks created https://github.com/moolitayer/awx-client-go/issues/22 so I won't forget

File: README.md (hunk 7 above)
Comment (moolitayer) [reply README.md]: I'll update, should be 1.0.5 https://groups.google.com/forum/#!searchin/awx-project/release%7Csort:date/awx-project/K80oF64C5EQ/1hWCKY5SAAAJ

File: README.md (hunk 14 above)
Comment (moolitayer) [reply README.md]: I didn't specify arguments in any of the methods above as I was trying to avoid complexity of `ExtraVar(name string, value interface{})` Instead I included an example for each. Do you think we should add a type in every example? (`Username(string)`)? just `...`? something else?

File: README.md [hunk 18]
```diff
@@ -11,6 +11,110 @@ go get github.com/golang/glog # Dependency
...
+```
+
+#### Launching a Job from a Template
+```go
+// Launch Job Template with id=8
+launchResource := connection.JobTemplates().Id(8).Launch()
+extraVars :=
+
+response, err := launchResource.Post()
+  .ExtraVars(map[string]string{"awx_environment": "staging"})
```
Comment (moolitayer) [inline README.md]: @zgalor is this about right?
BTW forgot to add `extraVar`,  will do

File: README.md [hunk 19]
```diff
@@ -11,6 +11,110 @@ go get github.com/golang/glog # Dependency
...
+  .Limit("master.example.com")
+  Send()
```
Comment (moolitayer) [inline README.md]: missing `.`

File: README.md (hunk 14 above)
Comment (jhernand) [reply README.md]: No, no need to add types to all examples, just use `Id` instead of `id`, and the ellipsis if you want to make it clear that this needs an argument.

File: README.md [hunk 20]
```diff
@@ -11,6 +11,110 @@ go get github.com/golang/glog # Dependency
...
+### import
+```go
+import 	"github.com/moolitayer/awx-client-go/awx"
+```
+
+### Creating a connection:
+```go
+// Uses the builder pattern:
+connection, err := awx.NewConnectionBuilder()
+  .Url("http://awx.example.com/api")          // Url is mandatory
```
Comment (jhernand) [inline README.md]: Note that the dots *must* be at the end of the lines, otherwise it will be a syntax error because the Go compiler automatically introduces a `;` at the end of the line if it doesn't end with that dot.

Comment (moolitayer) [reply README.md]: oh right thanks

File: README.md (hunk 18 above)
Comment (jhernand) [reply README.md]: If you are going to put the extra variables explicitly, then it is better to use the `ExtraVar` method, maybe multiple times:

```go
response, err := launchResource.Post().
  ExtraVar("awx_environment", "stating").
  ExtraVar("other", "variable").
  ...
  Limit("master.example.com").
  Send()
```

File: README.md [hunk 21]
```diff
@@ -11,6 +11,110 @@ go get github.com/golang/glog # Dependency
...
+
+```
+#### Retrieving resource by id
+Use `id()` on a resource list to get a single resource
+```go
+projectsResource := connection.Projects()
+
+// Get a resource managing a project with id=4
+getProjectsRequest := projectsResource.Get()
+projectResource := connection.Projects().Id(4)
```
Comment (jhernand) [inline README.md]: I think you don't need `projectsResource` and `getProjectsRequest` in this example.

File: README.md (hunk 18 above)
Comment (elad661) [reply README.md]: We should document both options, imo

File: README.md [hunk 22]
```diff
@@ -11,6 +11,109 @@ go get github.com/golang/glog # Dependency
...
+
+// Send the request to retrieve the project:
+getProjectResponse, err := projectResource.Get().Send()
+```
+
+#### Launching a Job from a Template
+```go
+// Launch Job Template with id=8
+launchResource := connection.JobTemplates().Id(8).Launch()
+extraVars :=
```
Comment (cben) [inline README.md]: incomplete statement.  I think can be dropped, you use map literal below.

File: README.md [hunk 23]
```diff
@@ -11,6 +11,109 @@ go get github.com/golang/glog # Dependency
...
+
+response, err := launchResource.Post().
+  ExtraVars(map[string]string{"awx_environment": "staging"}).
+  ExtraVar("instance", "example.com").
```
Comment (cben) [inline README.md]: IIUC using both ExtraVars then ExtraVar will mutate the map passed to ExtraVars.  Does this have to be documented?  Should ExtraVars make a copy?  (don't know what the Go etiquette is on mutatble args...)

**Review by cben — COMMENTED** [review]
doc looks really good!

File: README.md (hunk 23 above)
Comment (moolitayer) [reply README.md]: Do you see a draw back in changing the implementation? 

If no I would rather do that so that users don't have to care about it.

Comment (cben) [reply README.md]: :+1: my Python/Ruby instinct would be to change implementation to copy the passed map, I'm just too new to Go and don't know how things are done here :-)

**Review by jhernand — APPROVED** [review]

File: README.md (hunk 23 above)
Comment (moolitayer) [reply README.md]: I'm also not 100% sure, anyway I created https://github.com/moolitayer/awx-client-go/pull/29,
thanks
